src/core/public/application/capabilities/merge_capabilities.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Capabilities } from './capabilities_service';
+
+export function mergeCapabilities(sources: Array<Partial<Capabilities>>) {
+  const result: Capabilities = {
```

Comment:
If you wanted to avoid the imperative for loop, you could re-write as:

```ts
return sources.reduce((capabilities, source) => {
  Object.entries(source).forEach(([key, value]) => {
    capabilities[key] = { ...value, ...result[key] };
  });

  return capabilities;
}, {
  navLinks: {},
  management: {},
  catalogue: {},
} as Capabilities);
```